inception-external-search-elastic/src/main/java/de/tudarmstadt/ukp/inception/externalsearch/elastic/ElasticSearchProvider.java

```diff
@@ -126,9 +137,49 @@ public boolean isConnected()
                 }
             }
             if (hit.getHighlight() != null) {
+
+                // Highlights from elastic search are small sections of the document text
+                // with the keywords surrounded by the <em> tags. There are no offset information
+                // for the highlights or the keywords in the document text. There is a feature
+                // request for it (https://github.com/elastic/elasticsearch/issues/5736).
+                // Until this feature is implemented, we currently try to find the keywords offsets
+                // by finding the matching highlight in the document text, then the keywords offset
+                // within highlight using <em> tags.
+                String originalText = hit.get_source().getDoc().getText();
+
                 // There are highlights, set them in the result
-                List<String> highlights = new ArrayList<>();
-                highlights.add(hit.getHighlight().getDoctext().get(0));
+                List<ExternalSearchHighlight> highlights = new ArrayList<>();
+                for (String highlight : hit.getHighlight().getDoctext()) {
+
+                    // remove markers from the highlight
+                    String highlight_clean = highlight.replace(HIGHLIGHT_START_TAG, "")
+                        .replace(HIGHLIGHT_END_TAG, "");
+
+                    // find the matching highlight offset in the original text
+                    int highlight_start_index = originalText.indexOf(highlight_clean);
+
+                    // find offset to all keywords in the highlight
+                    // they are enclosed in <em> </em> tags in the highlight
+                    String highlightTemp = highlight;
+                    List<OffsetSpan> offsets = new ArrayList<>();
+                    while (highlightTemp.contains(HIGHLIGHT_START_TAG)) {
+                        int start = highlight_start_index +
+                            highlightTemp.indexOf(HIGHLIGHT_START_TAG);
+                        highlightTemp = highlightTemp.replaceFirst(HIGHLIGHT_START_TAG, "");
+                        int end = highlight_start_index +
+                            highlightTemp.indexOf(HIGHLIGHT_END_TAG);
+                        highlightTemp = highlightTemp.replaceFirst(HIGHLIGHT_END_TAG, "");
+                        offsets.add(new OffsetSpan(start, end));
+                    }
+
+                    if (!offsets.isEmpty()) {
```

Comment:
@rgabbard @shinyichen Here you fully skip highlights for which you cannot find a document position. I think this is too aggressive. I am planning to change this in https://github.com/inception-project/inception/pull/949 such that it is allowed to have `ExternalSearchHighlight`s without offset information. They can still render as highlights on the external search page and in the external search sidebar, even if the matches cannot be highlighted in the actual text. The reason is that there may be cases (e.g. the PubAnnotation connector) where I would like to show the highlights in the search results but at the time, I don't actually have access to the full text and therefore cannot calculate the offsets.

Reply:
@reckart : Your change seems reasonable to me.